components/raftstore-v2/src/operation/ready/apply_trace.rs

```diff
@@ -0,0 +1,603 @@
+///   interact with other peers will be traced.
+/// - support query the flushed progress without actually scanning raft engine,
+///   which is useful for cleaning up stale flush records.
+#[derive(Default)]
+pub struct ApplyTrace {
+    /// The modified indexes and flushed index of each data CF.
+    data_cfs: Box<[Progress; DATA_CFS_LEN]>,
+    /// The modified indexes and flushed index of raft CF.
+    ///
+    /// raft CF is a virtual CF that only used for recording apply index of
+    /// certain admin commands (like split/merge). So there is no flush at all.
+    /// The `flushed` field is advanced when the admin command doesn't need to
+    /// be replayed after restart. A write should be triggered to persist the
+    /// record.
+    admin: Progress,
+    /// Index that is issued to be written. It may not be truely persisted.
+    persisted_applied: u64,
+    /// `true` means the raft cf record should be persisted in next ready.
+    try_persist: bool,
+}
+
+impl ApplyTrace {
+    fn recover(region_id: u64, engine: &impl RaftEngine) -> Result<(Self, RegionLocalState)> {
+        let mut trace = ApplyTrace::default();
+        // Get all the recorded apply index from data CFs.
+        for (off, cf) in DATA_CFS.iter().enumerate() {
+            // There should be at least one record.
+            let i = engine.get_flushed_index(region_id, cf)?.unwrap();
+            trace.data_cfs[off].flushed = i;
+            trace.data_cfs[off].last_modified = i;
+        }
+        let i = engine.get_flushed_index(region_id, CF_RAFT)?.unwrap();
+        // Index of raft CF means all data before that must be persisted.
+        trace.admin.flushed = i;
+        trace.admin.last_modified = i;
+        trace.persisted_applied = i;
+        let applied_region_state = engine
+            .get_region_state(region_id, trace.admin.flushed)?
+            .unwrap();
+        Ok((trace, applied_region_state))
+    }
+
+    fn on_flush(&mut self, cf: &str, index: u64) {
+        let off = cf_offset(cf);
+        // Technically it should always be true.
+        if index > self.data_cfs[off].flushed {
+            self.data_cfs[off].flushed = index;
+        }
+    }
+
+    fn on_modify(&mut self, cf: &str, index: u64) {
+        let off = cf_offset(cf);
+        self.data_cfs[off].last_modified = index;
+    }
+
+    pub fn on_admin_flush(&mut self, index: u64) {
+        if index > self.admin.flushed {
+            self.admin.flushed = index;
+            self.try_persist = true;
+        }
+    }
+
+    pub fn on_admin_modify(&mut self, index: u64) {
+        self.admin.last_modified = index;
+    }
+
+    fn persisted_apply_index(&self) -> u64 {
+        self.admin.flushed
+    }
+
+    // All events before `mem_index` must be consumed before calling this function.
+    fn maybe_advance_admin_flushed(&mut self, mem_index: u64) {
+        if self.admin.flushed < self.admin.last_modified {
```

Comment:
I think it can be confusing sometimes for admin to reuse the progress struct. For normal progress it should have invariant `flushed <= last_modified`.

Reply:
No, the relation between `flushed` and `last_modified` can be arbitrary.